src/Orleans.Core/Runtime/QueuedExecutor.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Orleans.Runtime
+{
+    internal class QueuedExecutor : IExecutor
+    {
+        private QueueTrackingStatistic queueTracking;
+
+        private readonly BlockingCollection<QueueWorkItemCallback> workQueue = new BlockingCollection<QueueWorkItemCallback>();
+        private readonly CancellationTokenSource cancellationTokenSource;
+        private readonly bool drainAfterCancel;
+
+#if TRACK_DETAILED_STATS
+        internal protected ThreadTrackingStatistic threadTracking;
+#endif
+
+        public QueuedExecutor(string name, CancellationTokenSource cts, bool drainAfterCancel)
+        {
+            if (StatisticsCollector.CollectQueueStats)
+            {
+                queueTracking = new QueueTrackingStatistic(name);
+            }
+
+#if TRACK_DETAILED_STATS
+            if (StatisticsCollector.CollectThreadTimeTrackingStats)
+            {
+                threadTracking = new ThreadTrackingStatistic(Name);
+            }
+#endif
+            this.drainAfterCancel = drainAfterCancel;
+            cancellationTokenSource = cts;
+            cancellationTokenSource.Token.Register(() => workQueue.CompleteAdding());
+            new ThreadPerTaskExecutor(name).QueueWorkItem(_ => ProcessQueue());
+        }
+
+        public int WorkQueueCount => workQueue.Count;
+
+        public void QueueWorkItem(WaitCallback callback, object state = null)
+        {
+            var workItemCallback = new QueueWorkItemCallback(callback, state);
+
+#if TRACK_DETAILED_STATS
+            if (StatisticsCollector.CollectQueueStats)
+            {
+                queueTracking.OnEnQueueRequest(1, WorkQueueCount, workItemCallback);
+            }
+#endif
+
+            workQueue.Add(workItemCallback);
+        }
+
+        protected void ProcessQueue()
+        {
+#if TRACK_DETAILED_STATS
+            if (StatisticsCollector.CollectThreadTimeTrackingStats)
+            {
+                queueTracking.OnStartExecution();
+            }
+#endif
+            try
+            {
+                RunNonBatching();
+            }
+            finally
+            {
+#if TRACK_DETAILED_STATS
+                if (StatisticsCollector.CollectThreadTimeTrackingStats)
+                {
+                    queueTracking.OnStopExecution();
+                }
+#endif
+            }
+        }
+        
+        protected void RunNonBatching()
+        {
+            while (true)
+            {
+                if (!drainAfterCancel && cancellationTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                QueueWorkItemCallback workItem;
+                try
+                {
+                    workItem = workQueue.Take();
```

Comment:
The old `DrainQueue` logic used `TryTake`  whereas this uses `Take`. I suppose the idea is to make the queue throw `InvalidOperationException` when `Take` is called on the empty (and completed) queue. In Orleans today, I often see a bunch of those exceptions when a silo is being killed. Ultimately I think there's little that could be done about it except for maybe injecting one final no-op work item after the CTS has been canceled so threads get a chance to exit gracefully.

I'm ok with no-action here

Reply:
Added no-op action on cancel.